repose-aggregator/core/repose-core/src/main/scala/org/openrepose/powerfilter/ReposeRoutingServlet.scala

```diff
@@ -0,0 +1,362 @@
+
+  private val localCluster = new AtomicReference[Option[ReposeCluster]](None)
+  private val localNode = new AtomicReference[Option[Node]](None)
+  private val metricsService = new AtomicReference[Option[MetricsService]](Option(optMetricsService.orElse(null)))
+  private val defaultDestination = new AtomicReference[Option[Destination]](None)
+  private val destinations = new AtomicReference[Map[String, Destination]](Map.empty[String, Destination])
+  private val domains = new AtomicReference[Option[Clusters]](None) // TODO: Remove this when routing to a cluster is removed.
+  private var initialized = false
+
+  override def init(): Unit = {
+    logger.info("{}:{} -- Reticulating Splines - Initializing Repose Routing Servlet", clusterId, nodeId)
+    configurationService.subscribeTo(
+      SystemModelConfigurationFilename,
+      getClass.getResource("/META-INF/schema/system-model/system-model.xsd"),
+      this,
+      classOf[SystemModel]
+    )
+  }
+
+  override def destroy(): Unit = {
+    configurationService.unsubscribeFrom(SystemModelConfigurationFilename, this)
+    logger.info("{}:{} -- Obfuscated Quigley Matrix - Destroying Repose Routing Servlet", clusterId, nodeId)
+  }
+
+  override def configurationUpdated(configurationObject: SystemModel): Unit = {
+    //Set the current system model, and just update the nodes.
+    val interrogator = new SystemModelInterrogator(clusterId, nodeId)
+    localCluster.set(Option(interrogator.getLocalCluster(configurationObject).orElse(null)))
+    localNode.set(Option(interrogator.getLocalNode(configurationObject).orElse(null)))
+    defaultDestination.set(Option(interrogator.getDefaultDestination(configurationObject).orElse(null)))
+    destinations.set(localCluster.get().map(_.getDestinations) match {
+      case Some(dests) => getDestinations(dests.getEndpoint.asScala.toList) ++ getDestinations(dests.getTarget.asScala.toList)
+      case None => Map.empty[String, Destination]
+    })
+    domains.set(Option(new Clusters(configurationObject)))
+
+    if (logger.underlying.isDebugEnabled) {
+      localCluster.get().foreach { cluster =>
+        val clusterId: String = cluster.getId
+        //Build a list of the nodes in this cluster, just so we know what we're doing
+        val clusterNodes = cluster.getNodes.getNode.asScala.map(node => s"${node.getId}-${node.getHostname}")
+        logger.debug("{}:{} - Nodes for this router: {}", clusterId, nodeId, clusterNodes)
+        val destStrings = destinations.get()
+          .filter { case (_, dest) => dest.isInstanceOf[DestinationEndpoint] }
+          .map { case (_, endpoint) => s"${endpoint.getId}-${endpoint.asInstanceOf[DestinationEndpoint].getHostname}" }
+        logger.debug("{}:{} - Destinations for this router: {}", clusterId, nodeId, destStrings)
+      }
+    }
+    initialized = true
+
+    def getDestinations(destList: List[Destination]): Map[String, Destination] = {
+      destList.iterator.map(dest => dest.getId -> dest).toMap
```

Comment:
Why the `.iterator` call?

Also, since this code assumes destination IDs are unique (which should be the case, since it's enforced by the schema), you could do:
```
destList.groupBy(_.getId).mapValues(_.head)
```

Hm, and I wonder if we want to leave a comment stating that it is assumed that destination IDs must be unique.

Reply:
Smashed down to one readable line.